pkg/process/monitor/process_monitor.go

```diff
@@ -0,0 +1,247 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux
+// +build linux
+
+package monitor
+
+import (
+	"regexp"
+	"runtime"
+	"sync"
+
+	"github.com/vishvananda/netlink"
+
+	"github.com/DataDog/datadog-agent/pkg/network/config"
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/gopsutil/process"
+)
+
+const (
+	processMonitorMaxEvents = 2048
+)
+
+const (
+	STOPPED = iota
+	RUNNING
+)
+
+var (
+	processMonitor *ProcessMonitor
+)
+
+type ProcessMonitor struct {
+	m  sync.Mutex
+	wg sync.WaitGroup
+
+	state int
+
+	events chan netlink.ProcEvent
+	done   chan struct{}
+	errors chan error
+
+	procEventCallbacks map[ProcessEventType][]*ProcessCallback
+	callbackRunner     chan func()
+	callbackRunnerDone chan struct{}
+}
+
+type ProcessEventType int
+
+const (
+	EXEC ProcessEventType = iota
+	FORK
+	EXIT
+)
+
+type ProcessMetadataField int
+
+const (
+	ANY ProcessMetadataField = iota
+	NAME
+	MAPFILE
+)
+
+type ProcessCallback struct {
+	Event    ProcessEventType
+	Metadata ProcessMetadataField
+	Regex    *regexp.Regexp
+	Callback func(pid uint32)
+}
+
+func (pm *ProcessMonitor) enqueueCallback(callback *ProcessCallback, pid uint32) {
+	pm.callbackRunner <- func() { callback.Callback(pid) }
+}
+
+func (pm *ProcessMonitor) InitWithAllCurrentProcess() {
+	fn := func(pid int) error {
+		pm.m.Lock()
+		for _, c := range pm.procEventCallbacks[EXEC] {
+			pm.evalEXECCallback(c, uint32(pid))
+		}
+		pm.m.Unlock()
+		return nil
+	}
+
+	if err := util.WithAllProcs(util.HostProc(), fn); err != nil {
+		log.Errorf("process monitor init, scanning all process failed %s", err)
+	}
+
+	pm.state = RUNNING
+}
+
+func (p *ProcessMonitor) evalEXECCallback(c *ProcessCallback, pid uint32) {
+	if c.Metadata == ANY {
+		p.enqueueCallback(c, pid)
+		return
+	}
+
+	proc, err := process.NewProcess(int32(pid))
+	if err != nil {
+		log.Errorf("process %d parsing failed %s", pid, err)
+		return
+	}
+	switch c.Metadata {
+	case NAME:
+		pname, err := proc.Name()
+		if err != nil {
+			log.Errorf("process %d name parsing failed %s", pid, err)
+			return
+		}
+		if c.Regex.MatchString(pname) {
+			p.enqueueCallback(c, pid)
+		}
+	case MAPFILE:
+		mmaps, err := proc.MemoryMaps(true)
+		if err != nil {
+			log.Errorf("process %d/smaps parsing failed %s", pid, err)
+			return
+		}
+		if mmaps == nil {
+			return
+		}
+		for _, mmap := range *mmaps {
+			if c.Regex.MatchString(mmap.Path) {
+				p.enqueueCallback(c, pid)
+			}
+		}
```

Comment:
shouldn't we stop after the first `p.enqueueCallback(c, pid)`?

Reply:
a regex can match multiple path

Reply:
meanwhile you're right, if we match once or multiple time we should call the callback only once, it's on the user side to register and manage his regex